src/Suave/Logging/LogLine.fs

```diff
@@ -13,23 +15,45 @@ type LogLine =
     level         : LogLevel
     /// the source of the log line, e.g. 'ModuleName.FunctionName'
     path          : string
-    /// the message that the application wants to log
-    message       : string
+    /// the template message that the application wants to log
+    template      : string
+    /// a function which can format the log template message. When the
+    /// message is pre-formatted, this is None. When the log message 
+    /// contains structured data, this function can be used to render
+    /// the message template with the data.
+    formatter     : LogTemplateFormatter option
+    /// any key-value based data to log
+    data          : Map<string, obj>
     /// an optional exception
     ``exception`` : exn option
     /// timestamp when this log line was created
     tsUTCTicks    : int64 }
-    
+
+    /// Formats the log line as a string
+    member x.formatMessage () =
+        match x.formatter with
+        | None -> x.template
+        | Some fmter -> fmter x.template x.GetDataByName
+
+    /// Gets a data value
+    member x.GetDataByName name = Option.toObj (x.data.TryFind name)
+
 [<CompilationRepresentation(CompilationRepresentationFlags.ModuleSuffix)>]
 module LogLine =
   let mk path level trace ex message =
-    { message       = message
+    { template      = message
+      formatter     = None
+      data          = Map.empty
       level         = level
       path          = path
       ``exception`` = ex
       trace         = trace
       tsUTCTicks  = Globals.utcNow().Ticks }
 
+  /// assists with type inference, avoiding ugly 'box' calls for each
+  /// value item when constructing the log data map 
+  let mapLogData (data: (string*obj) list) = Map.ofList data
```

Comment:
Firstly, I am [not proud of that method](https://twitter.com/adamchester/status/632693196615610368), however I'd also like to point out, it looks a lot nicer if you view it like [this](https://twitter.com/hmemcpy/status/632805273380372480/photo/1):
![cmgtctrxaaapxad png-large](https://cloud.githubusercontent.com/assets/570470/9515536/de5417c4-4ce5-11e5-898e-a071955f3634.png)

Yes we can definitely do better (e.g. remove need for reflection) if we make the log message creator figure it all out up front, however I would have to weigh the cost/benefit to know for sure :)

I am not (yet, at least) proposing bringing anything from 'MessageTemplates' to Suave. Having said that, I *am* quite a fan of the Serilog approach to structured log messages, but I'm trying to keep an open mind right now, and keep myself open to alternative approaches for Suave.

Serilog and MessageTemplates goes to great lengths to ensure the captured structures are immediately copied to immutable structures, so that once rendering takes place, there is no risk of anything having been changed. Critically, it does this in a way that puts *minimal* burden on the log message writer/creator. See [here](https://github.com/serilog/serilog/wiki/Structured-Data).

IMO, it should remain dead simple to construct the log message template for a generic logging library. However, in the context of Suave, I'd love to explore other possible approaches too, which may push some of the burden back onto the Suave log message creator, in order to reduce complexity of capturing the logged messages. That burden may be so small, that I end up totally convinced it's not worth a message template library; actually this is why I haven't really proposed it yet.